pkg/virt-operator/resource/generate/components/secrets.go

```diff
@@ -273,7 +273,21 @@ func NextRotationDeadline(cert *tls.Certificate, ca *tls.Certificate, duration *
 	}
 
 	notAfter := cert.Leaf.NotAfter
-	deadline := notAfter.Add(-time.Duration(float64(duration.Duration) * 0.2))
+	caDeadline := notAfter
+	deadline := cert.Leaf.NotAfter.Add(-renewBefore.Duration)
+
+	if ca != nil {
+		caDeadline = ca.Leaf.NotAfter
+		if caRenewBefore != nil {
+			// Set cert rotation for the middle of the period of time when CA's overlap
+			caDeadline = notAfter.Add(-time.Duration(float64(caRenewBefore.Duration) * 0.5))
```

Comment:
@stu-gott I can't wrap my head around this :/

Why do we need to do this calculation? I think I'd expect us to act normally in this case.
If the CA duration is longer than the cert duration, why can't we simply start renewing `renewBefore` time before the cert expires?

Reply:
We compare the two candidate deadlines in the next step. and we take whichever one comes first. If the CA expires before the cert does then there's going to be a period of time where the cert hasn't rotated yet, but is invalid--leading to occasional TLS errors if API calls are made.

Aiming for the middle of the renewal overlap period means:
1) the new CA will have been rolled out by then so when we renew we'll use the new CA
2) there's still some time left on the old CA so we can avoid race conditions where there's still a miniscule window of time that we could try to use the old cert right at the expiry.

Reply:
If you're asking about the math itself, renewBefore is the time before the end of a cert's lifetime, so we need to subtract this from the "notAfter" (the actual validity deadline).

As per above comment, I wanted to aim for the middle of the overlap. if we take the exact deadline there's a race condition where the old now-invalid cert could be used at the last instant.

If we renew exactly at "notAfter minus renewBefore" (in other words, the exact moment a new CA will be calculated. we used to call this overlap interval), there's a race condition where we could renew an instant too early and create yet another cert with the old CA.

I split the difference to avoid both problems.

Reply:
Ok, thanks. I miss understood that this calculation is needed only when the CA expires before the cert.
Could this be that even if the CA expires after the certificate, its renewBefore/2 is long enough to be before the deadline?
Perhaps we should explicitly calculate this when the CA expires before the cert.